File: src/keybindings.ts
import type { Keybinding, KeybindingConfig, KeybindingPreset } from "./types";

export type Platform = "mac" | "other";

export interface ActionDefinition {
  action: string;
  label: string;
  context: string;
  intellij: string;
  vscode: string;
  vim: string;
  emacs: string;
}

export const ACTION_DEFINITIONS: ActionDefinition[] = [
  {
    action: "command_palette",
    label: "Command Palette",
    context: "global",
    intellij: "Ctrl+Shift+A / Cmd+Shift+A",
    vscode: "Ctrl+Shift+P / Cmd+Shift+P",
    vim: ":",
    emacs: "Alt+X",
  },
  {
    action: "new_terminal_tab",
    label: "New Terminal Tab",
    context: "terminal",
    intellij: "Alt+F12",
    vscode: "Ctrl+`",
    vim: ":term",
    emacs: "Ctrl+Alt+T",
  },
  { action: "close_tab", label: "Close Tab", context: "global", intellij: "Ctrl+F4", vscode: "Ctrl+W", vim: ":q", emacs: "Ctrl+X K" },
  { action: "split_horizontal", label: "Split Horizontal", context: "terminal", intellij: "Ctrl+Shift+H", vscode: "Ctrl+\\", vim: ":sp", emacs: "Ctrl+X 2" },
  { action: "split_vertical", label: "Split Vertical", context: "terminal", intellij: "Ctrl+Shift+V", vscode: "Ctrl+Shift+\\", vim: ":vsp", emacs: "Ctrl+X 3" },
  { action: "toggle_chat", label: "Toggle Chat", context: "global", intellij: "Alt+C", vscode: "Ctrl+Shift+I", vim: "<leader>c", emacs: "Ctrl+C C" },
  { action: "toggle_git_panel", label: "Toggle Git Panel", context: "global", intellij: "Alt+9", vscode: "Ctrl+Shift+G", vim: "<leader>g", emacs: "Ctrl+C G" },
  { action: "toggle_marketplace", label: "Toggle Marketplace", context: "global", intellij: "Alt+M", vscode: "Ctrl+Shift+M", vim: "<leader>m", emacs: "Ctrl+C M" },
  { action: "toggle_kanban", label: "Toggle Kanban", context: "global", intellij: "Alt+K", vscode: "Ctrl+Shift+K", vim: "<leader>k", emacs: "Ctrl+C K" },
  { action: "send_message", label: "Send Message", context: "chat", intellij: "Ctrl+Enter", vscode: "Ctrl+Enter", vim: "<CR>", emacs: "Ctrl+Enter" },
  { action: "cancel_agent", label: "Cancel Agent", context: "chat", intellij: "Ctrl+C", vscode: "Ctrl+C", vim: "<Esc>", emacs: "Ctrl+G" },
  { action: "focus_terminal", label: "Focus Terminal", context: "global", intellij: "Alt+F12", vscode: "Ctrl+`", vim: "<leader>t", emacs: "Ctrl+Alt+T" },
  { action: "focus_chat", label: "Focus Chat", context: "global", intellij: "Alt+C", vscode: "Ctrl+L", vim: "<leader>c", emacs: "Ctrl+X O" },
  { action: "next_tab", label: "Next Tab", context: "global", intellij: "Alt+Right", vscode: "Ctrl+Tab", vim: "gt", emacs: "Ctrl+PageDown" },
  { action: "prev_tab", label: "Prev Tab", context: "global", intellij: "Alt+Left", vscode: "Ctrl+Shift+Tab", vim: "gT", emacs: "Ctrl+PageUp" },
  { action: "search_files", label: "Search Files", context: "global", intellij: "Ctrl+Shift+F", vscode: "Ctrl+Shift+F", vim: "/", emacs: "Ctrl+S" },
  { action: "quick_open", label: "Quick Open", context: "global", intellij: "Ctrl+Shift+N", vscode: "Ctrl+P", vim: ":e", emacs: "Ctrl+X Ctrl+F" },
  { action: "settings", label: "Settings", context: "global", intellij: "Ctrl+Alt+S", vscode: "Ctrl+,", vim: ":set", emacs: "Ctrl+H V" },
  { action: "switch_mode", label: "Switch Mode", context: "global", intellij: "Ctrl+Shift+M", vscode: "Ctrl+Shift+M", vim: ":mode", emacs: "Ctrl+C Ctrl+M" },
  { action: "checkpoint", label: "Checkpoint", context: "global", intellij: "Ctrl+S", vscode: "Ctrl+S", vim: ":w", emacs: "Ctrl+X Ctrl+S" },
];

export function detectPlatform(value?: string): Platform {
  const platform = value ?? (typeof navigator === "undefined" ? "" : navigator.platform);
  return /mac|iphone|ipad|ipod/i.test(platform) ? "mac" : "other";
}

export function resolvePlatformShortcut(keys: string, platform: Platform): string {
  const normalized = keys.trim();
  if (!normalized.includes("/")) return normalized;
  const candidates = normalized.split("/").map((item) => item.trim());
  const prefer = platform === "mac" ? "cmd+" : "ctrl+";
  return candidates.find((item) => item.toLowerCase().includes(prefer)) ?? candidates[0] ?? normalized;
}

export function getPresetBindings(preset: KeybindingPreset): Keybinding[] {
  if (preset === "custom") {
    return [];
  }

  const key = preset as Exclude<KeybindingPreset, "custom">;
  return ACTION_DEFINITIONS.map((definition) => ({
    action: definition.action,
    keys: definition[key],
    context: definition.context,
  }));
}

export function mergeKeybindings(config: KeybindingConfig): Keybinding[] {
  const merged = [...getPresetBindings(config.preset)];
  for (const override of config.overrides) {
    const existingIndex = merged.findIndex(
      (binding) => binding.action === override.action && (binding.context ?? "global") === (override.context ?? "global")
    );
    if (existingIndex >= 0) {
      merged[existingIndex] = override;
    } else {
      merged.push(override);
    }
  }
  return merged;
}

export function normalizeCombo(combo: string): string {
  const parts = combo
    .split("+")
    .map((part) => part.trim())
    .filter(Boolean)
    .map(normalizeToken);

  const modifiers = ["Cmd", "Ctrl", "Alt", "Shift"];
  const ordered = [
    ...modifiers.filter((modifier) => parts.includes(modifier)),
    ...parts.filter((part) => !modifiers.includes(part)),
  ];
  return ordered.join("+");
}

function normalizeToken(token: string): string {
  const value = token.toLowerCase();
  if (["meta", "cmd", "command", "⌘"].includes(value)) return "Cmd";
  if (["control", "ctrl", "^"].includes(value)) return "Ctrl";
  if (["option", "alt", "⌥"].includes(value)) return "Alt";
  if (["shift", "⇧"].includes(value)) return "Shift";
  if (["escape", "esc"].includes(value)) return "Escape";
  if (value === "return") return "Enter";
  if (value === "space") return "Space";
  if (value === "arrowleft") return "Left";
  if (value === "arrowright") return "Right";
  if (value === "arrowup") return "Up";
  if (value === "arrowdown") return "Down";
  if (token.length === 1) return token.toUpperCase();
  return token;
}

function keyFromEvent(event: KeyboardEvent): string {
  if (event.key === " ") return "Space";
  if (event.key === "ArrowLeft") return "Left";
  if (event.key === "ArrowRight") return "Right";
  if (event.key === "ArrowUp") return "Up";
  if (event.key === "ArrowDown") return "Down";
  if (event.key.length === 1) return event.key.toUpperCase();
  return normalizeToken(event.key);
}

export function comboFromEvent(event: KeyboardEvent, platform: Platform): string {
  const combo: string[] = [];
  if (platform === "mac" ? event.metaKey : event.ctrlKey) combo.push(platform === "mac" ? "Cmd" : "Ctrl");
  if (platform === "mac" ? event.ctrlKey : event.metaKey) combo.push(platform === "mac" ? "Ctrl" : "Cmd");
  if (event.altKey) combo.push("Alt");
  if (event.shiftKey) combo.push("Shift");
  combo.push(keyFromEvent(event));
  return normalizeCombo(combo.join("+"));
}

function normalizeVimLiteral(shortcut: string): string {
  if (shortcut === "<CR>") return "Enter";
  if (shortcut === "<Esc>") return "Escape";
  return shortcut;
}

export function isMatch(event: KeyboardEvent, binding: Keybinding, platform: Platform): boolean {
  if (event.repeat || event.isComposing) return false;

  const resolved = resolvePlatformShortcut(binding.keys, platform);
  if (resolved.includes(" ") || resolved.includes("<leader>") || (/^:[a-z]/i.test(resolved) && resolved !== ":")) {
    return false;
  }

  const shortcut = normalizeVimLiteral(resolved);
  const normalizedShortcut = normalizeCombo(shortcut);
  const eventCombo = comboFromEvent(event, platform);
  return normalizedShortcut === eventCombo;
}

export function shouldPreventDefault(action: string): boolean {
  return [
    "command_palette",
    "close_tab",
    "search_files",
    "quick_open",
    "settings",
    "checkpoint",
    "new_terminal_tab",
  ].includes(action);
}

File: crates/caduceus-core/src/keybindings.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum KeybindingPreset {
    #[default]
    IntelliJ,
    VSCode,
    Vim,
    Emacs,
    Custom,
}

impl KeybindingPreset {
    pub fn all() -> Vec<Self> {
        vec![
            Self::IntelliJ,
            Self::VSCode,
            Self::Vim,
            Self::Emacs,
            Self::Custom,
        ]
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Keybinding {
    pub action: String,
    pub keys: String,
    pub context: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub struct KeybindingConfig {
    pub preset: KeybindingPreset,
    pub overrides: Vec<Keybinding>,
}

impl KeybindingConfig {
    pub fn preset_bindings(preset: KeybindingPreset) -> Vec<Keybinding> {
        match preset {
            KeybindingPreset::IntelliJ => vec![
                keybinding("command_palette", "Ctrl+Shift+A / Cmd+Shift+A", "global"),
                keybinding("new_terminal_tab", "Alt+F12", "terminal"),
                keybinding("close_tab", "Ctrl+F4", "global"),
                keybinding("split_horizontal", "Ctrl+Shift+H", "terminal"),
                keybinding("split_vertical", "Ctrl+Shift+V", "terminal"),
                keybinding("toggle_chat", "Alt+C", "global"),
                keybinding("toggle_git_panel", "Alt+9", "global"),
                keybinding("toggle_marketplace", "Alt+M", "global"),
                keybinding("toggle_kanban", "Alt+K", "global"),
                keybinding("send_message", "Ctrl+Enter", "chat"),
                keybinding("cancel_agent", "Ctrl+C", "chat"),
                keybinding("focus_terminal", "Alt+F12", "global"),
                keybinding("focus_chat", "Alt+C", "global"),
                keybinding("next_tab", "Alt+Right", "global"),
                keybinding("prev_tab", "Alt+Left", "global"),
                keybinding("search_files", "Ctrl+Shift+F", "global"),
                keybinding("quick_open", "Ctrl+Shift+N", "global"),
                keybinding("settings", "Ctrl+Alt+S", "global"),
                keybinding("switch_mode", "Ctrl+Shift+M", "global"),
                keybinding("checkpoint", "Ctrl+S", "global"),
            ],
            KeybindingPreset::VSCode => vec![
                keybinding("command_palette", "Ctrl+Shift+P / Cmd+Shift+P", "global"),
                keybinding("new_terminal_tab", "Ctrl+`", "terminal"),
                keybinding("close_tab", "Ctrl+W", "global"),
                keybinding("split_horizontal", "Ctrl+\\", "terminal"),
                keybinding("split_vertical", "Ctrl+Shift+\\", "terminal"),
                keybinding("toggle_chat", "Ctrl+Shift+I", "global"),
                keybinding("toggle_git_panel", "Ctrl+Shift+G", "global"),
                keybinding("toggle_marketplace", "Ctrl+Shift+M", "global"),
                keybinding("toggle_kanban", "Ctrl+Shift+K", "global"),
                keybinding("send_message", "Ctrl+Enter", "chat"),
                keybinding("cancel_agent", "Ctrl+C", "chat"),
                keybinding("focus_terminal", "Ctrl+`", "global"),
                keybinding("focus_chat", "Ctrl+L", "global"),
                keybinding("next_tab", "Ctrl+Tab", "global"),
                keybinding("prev_tab", "Ctrl+Shift+Tab", "global"),
                keybinding("search_files", "Ctrl+Shift+F", "global"),
                keybinding("quick_open", "Ctrl+P", "global"),
                keybinding("settings", "Ctrl+,", "global"),
                keybinding("switch_mode", "Ctrl+Shift+M", "global"),
                keybinding("checkpoint", "Ctrl+S", "global"),
            ],
            KeybindingPreset::Vim => vec![
                keybinding("command_palette", ":", "global"),
                keybinding("new_terminal_tab", ":term", "terminal"),
                keybinding("close_tab", ":q", "global"),
                keybinding("split_horizontal", ":sp", "terminal"),
                keybinding("split_vertical", ":vsp", "terminal"),
                keybinding("toggle_chat", "<leader>c", "global"),
                keybinding("toggle_git_panel", "<leader>g", "global"),
                keybinding("toggle_marketplace", "<leader>m", "global"),
                keybinding("toggle_kanban", "<leader>k", "global"),
                keybinding("send_message", "<CR>", "chat"),
                keybinding("cancel_agent", "<Esc>", "chat"),
                keybinding("focus_terminal", "<leader>t", "global"),
                keybinding("focus_chat", "<leader>c", "global"),
                keybinding("next_tab", "gt", "global"),
                keybinding("prev_tab", "gT", "global"),
                keybinding("search_files", "/", "global"),
                keybinding("quick_open", ":e", "global"),
                keybinding("settings", ":set", "global"),
                keybinding("switch_mode", ":mode", "global"),
                keybinding("checkpoint", ":w", "global"),
            ],
            KeybindingPreset::Emacs => vec![
                keybinding("command_palette", "Alt+X", "global"),
                keybinding("new_terminal_tab", "Ctrl+Alt+T", "terminal"),
                keybinding("close_tab", "Ctrl+X K", "global"),
                keybinding("split_horizontal", "Ctrl+X 2", "terminal"),
                keybinding("split_vertical", "Ctrl+X 3", "terminal"),
                keybinding("toggle_chat", "Ctrl+C C", "global"),
                keybinding("toggle_git_panel", "Ctrl+C G", "global"),
                keybinding("toggle_marketplace", "Ctrl+C M", "global"),
                keybinding("toggle_kanban", "Ctrl+C K", "global"),
                keybinding("send_message", "Ctrl+Enter", "chat"),
                keybinding("cancel_agent", "Ctrl+G", "chat"),
                keybinding("focus_terminal", "Ctrl+Alt+T", "global"),
                keybinding("focus_chat", "Ctrl+X O", "global"),
                keybinding("next_tab", "Ctrl+PageDown", "global"),
                keybinding("prev_tab", "Ctrl+PageUp", "global"),
                keybinding("search_files", "Ctrl+S", "global"),
                keybinding("quick_open", "Ctrl+X Ctrl+F", "global"),
                keybinding("settings", "Ctrl+H V", "global"),
                keybinding("switch_mode", "Ctrl+C Ctrl+M", "global"),
                keybinding("checkpoint", "Ctrl+X Ctrl+S", "global"),
            ],
            KeybindingPreset::Custom => Vec::new(),
        }
    }

    pub fn effective_bindings(&self) -> Vec<Keybinding> {
        let mut bindings = Self::preset_bindings(self.preset);
        for override_binding in &self.overrides {
            if let Some(existing) = bindings.iter_mut().find(|binding| {
                binding.action == override_binding.action
                    && binding.context == override_binding.context
            }) {
                *existing = override_binding.clone();
            } else {
                bindings.push(override_binding.clone());
            }
        }
        bindings
    }
}

pub fn resolve_platform_shortcut(keys: &str, is_macos: bool) -> String {
    if !keys.contains('/') {
        return keys.trim().to_string();
    }

    let mut selected = keys.trim().to_string();
    for part in keys.split('/') {
        let normalized = part.trim();
        if is_macos && normalized.to_lowercase().contains("cmd+") {
            selected = normalized.to_string();
            break;
        }
        if !is_macos && normalized.to_lowercase().contains("ctrl+") {
            selected = normalized.to_string();
            break;
        }
    }
    selected
}

fn keybinding(action: &str, keys: &str, context: &str) -> Keybinding {
    Keybinding {
        action: action.to_string(),
        keys: keys.to_string(),
        context: Some(context.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_preset_is_intellij() {
        assert_eq!(
            KeybindingConfig::default().preset,
            KeybindingPreset::IntelliJ
        );
    }

    #[test]
    fn intellij_preset_contains_expected_command_palette_binding() {
        let bindings = KeybindingConfig::preset_bindings(KeybindingPreset::IntelliJ);
        let palette = bindings
            .iter()
            .find(|binding| binding.action == "command_palette")
            .expect("command_palette binding");
        assert_eq!(palette.keys, "Ctrl+Shift+A / Cmd+Shift+A");
    }

    #[test]
    fn vscode_preset_contains_expected_quick_open_binding() {
        let bindings = KeybindingConfig::preset_bindings(KeybindingPreset::VSCode);
        let quick_open = bindings
            .iter()
            .find(|binding| binding.action == "quick_open")
            .expect("quick_open binding");
        assert_eq!(quick_open.keys, "Ctrl+P");
    }

    #[test]
    fn vim_preset_contains_expected_mode_binding() {
        let bindings = KeybindingConfig::preset_bindings(KeybindingPreset::Vim);
        let switch_mode = bindings
            .iter()
            .find(|binding| binding.action == "switch_mode")
            .expect("switch_mode binding");
        assert_eq!(switch_mode.keys, ":mode");
    }

    #[test]
    fn override_replaces_existing_binding() {
        let config = KeybindingConfig {
            preset: KeybindingPreset::IntelliJ,
            overrides: vec![Keybinding {
                action: "command_palette".to_string(),
                keys: "Ctrl+K".to_string(),
                context: Some("global".to_string()),
            }],
        };

        let palette = config
            .effective_bindings()
            .into_iter()
            .find(|binding| binding.action == "command_palette")
            .expect("command_palette binding");

        assert_eq!(palette.keys, "Ctrl+K");
    }

    #[test]
    fn override_adds_new_action_binding() {
        let config = KeybindingConfig {
            preset: KeybindingPreset::VSCode,
            overrides: vec![Keybinding {
                action: "open_notifications".to_string(),
                keys: "Ctrl+Shift+N".to_string(),
                context: Some("global".to_string()),
            }],
        };

        let bindings = config.effective_bindings();
        assert!(bindings.iter().any(
            |binding| binding.action == "open_notifications" && binding.keys == "Ctrl+Shift+N"
        ));
    }

    #[test]
    fn platform_shortcut_resolves_to_cmd_on_macos() {
        let resolved = resolve_platform_shortcut("Ctrl+Shift+A / Cmd+Shift+A", true);
        assert_eq!(resolved, "Cmd+Shift+A");
    }

    #[test]
    fn platform_shortcut_resolves_to_ctrl_on_non_macos() {
        let resolved = resolve_platform_shortcut("Ctrl+Shift+A / Cmd+Shift+A", false);
        assert_eq!(resolved, "Ctrl+Shift+A");
    }
}
